bluefog/common/tensor_queue.cc

```diff
@@ -23,35 +22,107 @@ namespace bluefog {
 namespace common {
 
 // Add a TensorTableEntry as well as its message to the queue.
-Status TensorQueue::AddToTensorQueue(TensorTableEntry& e) {
+Status TensorQueue::AddToTensorQueue(TensorTableEntry& e, Request& message) {
   std::lock_guard<std::mutex> guard(mutex_);
-  message_queue_.push(std::move(e));
+  if (tensor_table_.find(e.tensor_name) != tensor_table_.end()) {
+    return DUPLICATE_NAME_ERROR;
+  }
+  const std::string& name = message.tensor_name();
+  tensor_table_.emplace(name, std::move(e));
+  message_queue_.push(message);
   return Status::OK();
 }
 
 // Put callbacks for each tensor in the callback buffer and clear tensor queue
 void TensorQueue::FinalizeTensorQueue(
     std::vector<StatusCallback>& callbacks_buffer) {
   std::lock_guard<std::mutex> guard(mutex_);
+  for (auto& e : tensor_table_) {
+    callbacks_buffer.emplace_back(e.second.callback);
+  }
+  tensor_table_.clear();
   while (!message_queue_.empty()) {
-    TensorTableEntry message = message_queue_.front();
-    BFLOG(TRACE) << "Message " << message.tensor_name << " is still in the queue after shut down.";
     message_queue_.pop();
-    callbacks_buffer.emplace_back(message.callback);
   }
 }
 
-// Pop out the front messages from the queue
-TensorTableEntry TensorQueue::PopMessagesFromQueue() {
+// Parse tensor names from response and generate a vector of corresponding
+// tensor entries.
+void TensorQueue::GetTensorEntriesFromResponse(
+    const Response& response, std::vector<TensorTableEntry>& entries) {
+  // Reserve to save re-allocation costs, as we know the size before.
+  // entries may not be empty due to win_ops is processed at first.
+  entries.reserve(entries.size() + response.tensor_names().size());
+  {
+    // Lock on the tensor table.
+    std::lock_guard<std::mutex> guard(mutex_);
+    for (auto& name : response.tensor_names()) {
+      auto iter = tensor_table_.find(name);
+      assert(iter != tensor_table_.end());
+
+      assert(response.response_type() == Response::ALLREDUCE ||
+             response.response_type() == Response::ALLGATHER ||
+             response.response_type() == Response::BROADCAST ||
+             response.response_type() == Response::NEIGHBOR_ALLGATHER ||
+             response.response_type() == Response::NEIGHBOR_ALLREDUCE ||
+             response.response_type() == Response::WIN_CREATE ||
+             response.response_type() == Response::WIN_FREE ||
+             response.response_type() == Response::ERROR);
+
+      if (response.response_type() == Response::ERROR) {
+        auto& e = iter->second;
+        e.callback(Status::PreconditionError(response.error_message()));
+      } else {
+        entries.push_back(std::move(iter->second));
+      }
+
+      // Clear the tensor table of this tensor.
+      tensor_table_.erase(iter);
+    }
+  }
+}
+
+// It should be used for no-coordinate request operator only.
```

Comment:
Shall we check the request type here for safety?

Reply:
When I set skip negotiation, all types of requests may pass in